controller/src/main/java/io/pravega/controller/store/index/PravegaTablesHostIndex.java

```diff
@@ -0,0 +1,143 @@
+        CompletableFuture<Void> createTableFuture; 
+        if (!Optional.ofNullable(hostIdMap.get(hostId)).orElse(false)) {
+            createTableFuture = storeHelper.createTable(SYSTEM_SCOPE, hostsTable)
+                    .thenCompose(x -> storeHelper.addNewEntryIfAbsent(SYSTEM_SCOPE, hostsTable, hostId, new byte[0]))
+                    .thenCompose(x -> storeHelper.createTable(SYSTEM_SCOPE, hostEntityTable))
+                                           
+                    .thenAccept(x -> hostIdMap.put(hostId, true));
+        } else {
+            createTableFuture = CompletableFuture.completedFuture(null);
+        }
+        
+        return createTableFuture.thenCompose(x -> {
+            return Futures.toVoid(storeHelper.addNewEntryIfAbsent(SYSTEM_SCOPE, hostEntityTable, entity, entityData));
+        });
+    }
+
+    private String getHostEntityTableName(String hostId) {
+        return String.format(HOST_TABLE_FORMAT, indexName, hostId);
+    }
+
+    @Override
+    public CompletableFuture<byte[]> getEntityData(String hostId, String entity) {
+        Preconditions.checkNotNull(hostId);
+        Preconditions.checkNotNull(entity);
+        String table = getHostEntityTableName(hostId);
+        return storeHelper.getEntry(SYSTEM_SCOPE, table, entity)
+                .thenApply(Data::getData);
+    }
+
+    @Override
+    public CompletableFuture<Void> removeEntity(final String hostId, final String entity, final boolean deleteEmptyHost) {
+        Preconditions.checkNotNull(hostId);
+        Preconditions.checkNotNull(entity);
+        String table = getHostEntityTableName(hostId);
+        return storeHelper.removeEntry(SYSTEM_SCOPE, table, entity)
+                .thenCompose(e -> {
+                    if (deleteEmptyHost) {
+                        return removeHost(hostId);
+                    } else {
+                        return CompletableFuture.completedFuture(null);
+                    }
+                });
+    }
+
+    @Override
+    public CompletableFuture<Void> removeHost(final String hostId) {
+        Preconditions.checkNotNull(hostId);
+        String table = getHostEntityTableName(hostId);
+        return storeHelper.deleteTable(SYSTEM_SCOPE, table, true)
+                .thenCompose(deleted -> {
+                    if (deleted) {
+                        return storeHelper.removeEntry(SYSTEM_SCOPE, hostsTable, hostId);     
+                    } else {
+                        return CompletableFuture.completedFuture(null);
+                    }
+                });
+    }
+
+    @Override
+    public CompletableFuture<List<String>> getEntities(final String hostId) {
+        Preconditions.checkNotNull(hostId);
+        String table = getHostEntityTableName(hostId);
+        List<String> entries = new LinkedList<>();
+        return storeHelper.getAllKeys(SYSTEM_SCOPE, table)
+                .forEachRemaining(entries::add, executor)
```

Comment:
This involves multiple threads and needs to be made thread safe. I think a better solution would be to implement a 'collect' method on async iterator.